tests/FsCodec.SystemTextJson.Tests/UnionConverterTests.fs

```diff
@@ -0,0 +1,361 @@
+
+type FsCheckGenerators =
+    static member CartId = Arb.generate |> Gen.map CartId |> Arb.fromGen
+    static member SkuId = Arb.generate |> Gen.map SkuId |> Arb.fromGen
+
+type DomainPropertyAttribute() =
+    inherit FsCheck.Xunit.PropertyAttribute(QuietOnSuccess = true, Arbitrary=[| typeof<FsCheckGenerators> |])
+
+let roundtripProperty ignoreNulls (profile : JsonSerializerOptions) value =
+    let serialized = JsonSerializer.Serialize(value, profile)
+    render ignoreNulls value =! serialized
+    let deserialized = JsonSerializer.Deserialize<_>(serialized, profile)
+    deserialized =! value
+
+let includeNullsProfile = Options.Create(ignoreNulls = false)
+[<DomainPropertyAttribute(MaxTest=1000)>]
+let ``UnionConverter ignoreNulls Profile roundtrip property test`` (x: TestDU) =
+    let ignoreNulls, profile = false, includeNullsProfile
+    profile.IgnoreNullValues =! false
+    roundtripProperty ignoreNulls profile x
+
+let defaultProfile = Options.Create ()
+[<DomainPropertyAttribute(MaxTest=1000)>]
+let ``UnionConverter opinionated Profile roundtrip property test`` (x: TestDU) =
+    let ignoreNulls, profile = false, defaultProfile
+    profile.IgnoreNullValues =! false
+    roundtripProperty ignoreNulls profile x
+
+//[<Fact>]
+//let ``Implementation ensures no internal errors escape (which would render a WebApi ModelState.Invalid)`` () =
+//    let s = JsonSerializer.CreateDefault()
+//    let mutable gotError = false
+//    s.Error.Add(fun _ -> gotError <- true)
+
+//    let dJson = """{"case":"CaseD","a":"hi"}"""
+//    use dReader = new StringReader(dJson)
+//    use dJsonReader = new JsonTextReader(dReader)
+//    let d = s.Deserialize<TestDU>(dJsonReader)
+
+//    test <@ (CaseD "hi") = d @>
+//    test <@ false = gotError @>
+
+module ``Unmatched case handling`` =
+    [<Fact>]
+    let ``UnionConverter by default throws on unknown cases`` () =
+        let options = Options.Create(UnionConverter ())
+        let aJson = """{"case":"CaseUnknown"}"""
+        let act () = JsonSerializer.Deserialize<TestDU>(aJson, options)
+
+        fun (e : System.InvalidOperationException) -> <@ -1 <> e.Message.IndexOf "No case defined for 'CaseUnknown', and no catchAllCase nominated" @>
+        |> raisesWith <@ act() @>
+
+    type DuWithCatchAll =
+    | Known
+    | Catchall
+
+    [<Fact>]
+    let ``UnionConverter supports a nominated catchall`` () =
+        let options = Options.Create(UnionConverter<DuWithCatchAll> ("case", "Catchall"))
+        let aJson = """{"case":"CaseUnknown"}"""
+        let a = JsonSerializer.Deserialize<DuWithCatchAll>(aJson, options)
+
+        test <@ Catchall = a @>
+
+    type DuWithMissingCatchAll =
+    | Known
+
+    [<Fact>]
+    let ``UnionConverter explains if nominated catchAll not found`` () =
+        let options = Options.Create(UnionConverter<DuWithMissingCatchAll> ("case", "CatchAllThatCantBeFound"))
```

Comment:
You can't pass parameters to JsonConverters via the attribute method any more, as far as I can tell. 
The workaround right now is to do what I'm doing here. If it's still important to have it settable via attributes we might need our own attribute, say `JsonUnionConverterOptionsAttribute` that would be used in tandem with the `JsonConverterAttribute`. (e.g. `[<JsonConverter(typeof<UnionConverter<MyDU>>); JsonUnionConverterOptions("case", "Catchall")>]`.)

Reply:
Good point; sounds like a practical solution...

Reply:
Had a good look about, couldn't see any other approach...

Reply:
Implemented in https://github.com/jet/FsCodec/pull/43/commits/4c78b79d1698620b7bd4310853772fdb640ec8ca